pkg/services/provisioning/dashboards/file_reader_test.go

```diff
@@ -101,55 +100,97 @@ func TestDashboardFileReader(t *testing.T) {
+					On("GetProvisionedDashboardData", "Default").
+					Return([]*models.DashboardProvisioning{}, nil)
+
+				fakeService.
+					On("SaveProvisionedDashboard", mock.Anything, mock.Anything).
+					Return(&models.Dashboard{}, nil).
+					Times(2)
+
 				err = reader.startWalkingDisk()
 				So(err, ShouldBeNil)
 
-				folders := 0
-				dashboards := 0
-
-				for _, i := range fakeService.inserted {
-					if i.Dashboard.IsFolder {
-						folders++
-					} else {
-						dashboards++
-					}
-				}
+				dash := &dashboards.SaveDashboardDTO{}
+				dash.Dashboard = models.NewDashboardFolder(cfg.Folder)
+				dash.Dashboard.IsFolder = true
+				dash.Overwrite = true
+				dash.OrgId = cfg.OrgId
 
-				So(folders, ShouldEqual, 1)
-				So(dashboards, ShouldEqual, 2)
+				fakeService.AssertExpectations(t)
 			})
 
+			// There was an issue with former test that it did not mock GetProvisionedDashboardData, which
+			// is the function that actually returns existing provisioned dashboards so it inserted a dashboard
+			// but not due to update but as a new dashboard
 			Convey("Can read default dashboard and replace old version in database", func() {
 				cfg.Options["path"] = oneDashboard
 
-				stat, _ := os.Stat(oneDashboard + "/dashboard1.json")
-
-				fakeService.getDashboard = append(fakeService.getDashboard, &models.Dashboard{
-					Updated: stat.ModTime().AddDate(0, 0, -1),
-					Slug:    "grafana",
-				})
-
-				reader, err := NewDashboardFileReader(cfg, logger)
-				So(err, ShouldBeNil)
+				dashboardPath := oneDashboard + "/dashboard1.json"
+				stat, _ := os.Stat(dashboardPath)
 
-				err = reader.startWalkingDisk()
+				absPath, err := filepath.Abs(dashboardPath)
 				So(err, ShouldBeNil)
 
-				So(len(fakeService.inserted), ShouldEqual, 1)
-			})
-
-			Convey("Overrides id from dashboard.json files", func() {
-				cfg.Options["path"] = containingId
+				// Fake older data in the DB
+				fakeService.
+					On("GetProvisionedDashboardData", "Default").
+					Return(
+						[]*models.DashboardProvisioning{
+							{
+								Id:          1,
+								DashboardId: 1,
+								Updated:     stat.ModTime().AddDate(0, 0, -1).Unix(),
+								ExternalId:  absPath,
+								CheckSum:    "",
+								Name:        "Default",
+							},
+						},
+						nil,
+					)
+
+				// And that dashboard is updated due to date difference
+				fakeService.
+					On("SaveProvisionedDashboard", mock.Anything, mock.Anything).
+					Return(&models.Dashboard{}, nil)
 
 				reader, err := NewDashboardFileReader(cfg, logger)
 				So(err, ShouldBeNil)
 
 				err = reader.startWalkingDisk()
 				So(err, ShouldBeNil)
 
-				So(len(fakeService.inserted), ShouldEqual, 1)
+				fakeService.AssertExpectations(t)
 			})
 
+			// Not sure what this should do, looking at the code we do not actually save the dashboard under the ID
+			// set in the json file, we strip it or use the one from dashboard_provisioning
+			//Convey("Overrides id from dashboard.json files", func() {
```

Comment:
Not sure what to do with this, maybe it was just forgotten? When looking at the code https://github.com/grafana/grafana/blob/master/pkg/services/provisioning/dashboards/file_reader.go#L175 it does not seem like we allow to save dashboard under specific Id so not sure if this should be fixed or just removed.